_posts/2017-08-18-Refactoring-with-reactJS-shared-state.md

```diff
@@ -0,0 +1,252 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import SuccessBanner from './successBanner.jsx';
+import FailureBanner from './failureBanner.jsx';
+import $ from 'jquery';
+import { createStore } from 'redux';
+import { Provider } from 'react-redux';
+import reducer from './reducers/index.jsx'
+
+// create the Redux store on the window to make accessible to whole application
+window.store = createStore(reducer);
+
+const componentRegistry = {
+    SuccessBanner,
+    FailureBanner
+};
+
+$(()=>{
+    const reactComponents = $('[data-react-component]');
+
+    $.each(reactComponents, function(index, obj) {
+        const props = $(obj).data();
+        const componentName = $(obj).data('reactComponent');
+        const Component = componentRegistry[componentName];
+
+        if (Component) {
+            // Render the component into the DOM
+            // wrapped in Provider to give it access to the Redux store
+            ReactDOM.render(
+                <Provider store={window.store}>
+                    <Component {...props} />
+                </Provider>,
+                obj
+            );
+        }
+    });
+});
+{% endhighlight %}
+
+Since we're interested in keeping track of the `FailureBanner`'s props in the application state, we wrote a simple reducer (that the store is created with). In this example, we assume that we only need to track the props for one `FailureBanner`. We store the `isShown` and `message` props as the application state. Any action dispatched with the type `SET_FAILURE_BANNER` updates the properties on the state.
+
+{% highlight javascript linenos %}
+const initialState = {
+    isShown: false,
+    message: ''
+};
+export default function(state = initialState, action) {
+    switch (action.type) {
+        case 'SET_FAILURE_BANNER':
+            return {
+                isShown: action.payload.isShown,
+                message: action.payload.message
+            };
+        default:
+            return state;
+    }
+}
+{% endhighlight %}
+
+Finally, we'll write an [action creator](https://daveceddia.com/redux-action-creators/) that creates the action object we want to dispatch when we want to change the Failure Banner state.
+
+{% highlight javascript linenos %}
+function setFailureBanner(isShown = false, message = '') {
+    return {
+        type: 'SET_FAILURE_BANNER',
+        payload: {
+            isShown,
+            message
+        }
+    };
+}
+{% endhighlight %}
+
+Now that we have a basic Redux store setup, we need to use the application state in our components.
+
+# Using the Redux State in React
+
+We'll address the most common pattern - using Redux with React. There are a lot of excellent resources that cover how to add Redux to your React application, including [this basic tutorial in the Redux docs](http://redux.js.org/docs/basics/). We trust that you're either familiar with the pattern or can read about it elsewhere, so we won't go into too much depth on the specifics.
```

Comment:
We made an effort to provide lightest seams possible to wedge in this implementation in order to preserve common uses of React and Redux. This means that all of the common implementation patterns and documentation do apply.